### PR TITLE
parser: add plan recreator statement

### PR DESCRIPTION
**Pull request by rebelice**
### What problem does this PR solve? 

This is a step in [tidb/issues/26325](https://github.com/pingcap/tidb/issues/26325)

### What is changed and how it works?

Add `PLAN RECREATOR` statement

### Check List 

Tests 

 - Unit test
 - Integration test

Code changes

 - Has exported function/method change

Side effects

 - Possible performance regression
 - Increased code complexity

Related changes

 - Need to update the documentation
 - Need to be included in the release note

### CODE REVIEW

File: parser.y [hunk 1]
```diff
@@ -13158,4 +13166,102 @@ RowStmt:
...
+		if $8 != nil {
+			x.OrderBy = $8.(*ast.OrderByClause)
+		}
+		if $9 != nil {
+			x.Limit = $9.(*ast.Limit)
+		}
+
+		$$ = x
+	}
+|	"PLAN" "RECREATOR" "DUMP" "EXPLAIN" "ANALYZE" "SLOW" "QUERY" WhereClauseOptional OrderByOptional SelectStmtLimitOpt
```
Comment (xhebox) [inline parser.y:13233]: Do you think `AnalyzeOptional` will better? Not a necessary change request.

**Review by xhebox — COMMENTED** [review]
Rest LGTM. BTW, the document is internal, and there is no corresponding RFC on github.

File: parser.y (hunk 1 above)
Comment (rebelice) [reply parser.y:13233]: Agree with you. I'll modify it later.

**Comment by rebelice** [comment]
> Rest LGTM. BTW, the document is internal, and there is no corresponding RFC on github.

Here is a description of this feature in https://github.com/pingcap/tidb/issues/26325. This is the new feature for Sprint 4, I'll add documents visible to everyone before this feature GA. Is this OK for you?

**Comment by xhebox** [comment]
@rebelice Sure.

File: parser.y (hunk 1 above)
Comment (rebelice) [reply parser.y:13237]: When I added `AnalyzeOptional`, I found that this abstraction seemed not necessary but added more work. I currently prefer not to modify this.

File: ast/misc.go [hunk 2]
```diff
@@ -343,6 +343,11 @@ func (n *PlanRecreatorStmt) Accept(v Visitor) (Node, bool) {
 		return n, false
 	}
 	n.Stmt = node.(StmtNode)
+	stmt, ok := n.Stmt.Accept(v)
```
Comment (xhebox) [inline ast/misc.go:346]: Accept twice?

Comment (rebelice) [reply ast/misc.go]: Sorry about that. Fixed.

**Review by xhebox — APPROVED** [review]

File: parser.y [hunk 3]
```diff
@@ -5582,6 +5586,7 @@ UnReservedKeyword:
 |	"DAY"
 |	"DEALLOCATE"
 |	"DO"
+|	"DUMP"
```
Comment (tiancaiamao) [inline parser.y:5589]: It's not in the UnReservedKeyword list:
https://dev.mysql.com/doc/refman/8.0/en/keywords.html

It should be NotKeyword?

File: parser.y [hunk 4]
```diff
@@ -5624,6 +5629,7 @@ UnReservedKeyword:
 |	"PROXY"
 |	"QUICK"
 |	"REBUILD"
+|	"RECREATOR"
```
Comment (tiancaiamao) [inline parser.y:5632]: ditto

File: parser.y [hunk 5]
```diff
@@ -5719,6 +5725,7 @@ UnReservedKeyword:
 |	"PROFILES"
 |	"MICROSECOND"
 |	"MINUTE"
+|	"PLAN"
```
Comment (tiancaiamao) [inline parser.y:5728]: ditto ... plan is also not a keyword

File: parser.y (hunk 3 above)
Comment (rebelice) [reply parser.y]: updated all

**Review by tiancaiamao — APPROVED** [review]

**Pull request merged**